### PR TITLE
Fetch statsgo

### PR DESCRIPTION
**Pull request by msleckman**
This PR is ready for review. 

**Added 3 targets:** 
`p1_selected_statsgo_sbid_children`: Pulls the specific child items from the Statsgo soil characteristics data release from Science base. Only interested in Texture attributes and Soil Layer attributes. _(intermediate target)_
`p1_download_statsgo_text_layer_attr`: file target downloading the statsgo zip files _(intermediate target)_
`p1_statsgo_soil_df`: reads in and scopes statsgo dataset + cbinds so that we have 1 target df object with total upstream (TOT), accumulated (ACC), and catchment (CAT)-level values. _(final fetch target)_

**Added 1 function:** 
`sb_read_filter_by_comids()` outputs `p1_statsgo_soil_df`. Built for statsgo but generalizable to any comid-level sb dataset that requires fetching and initial scoping (col selection and aoi row filtering). 

`p1_statsgo_soil_df` output:
```
   COMID ACC_KFACT ACC_KFACT_UP ACC_NO10AVE ACC_NO4AVE ACC_WTDEP CAT_KFACT CAT_KFACT_UP CAT_NO10AVE CAT_NO4AVE CAT_WTDEP TOT_KFACT TOT_KFACT_UP TOT_NO10AVE
1 1748535      0.28         0.24       49.80      57.25      4.89      0.28         0.24       49.80      57.25      4.89      0.28         0.24       49.80
2 1748537      0.28         0.24       51.47      58.49      4.41      0.28         0.24       51.47      58.49      4.41      0.28         0.24       51.47
3 1748539      0.28         0.24       52.71      59.68      3.77      0.28         0.24       52.71      59.68      3.77      0.28         0.24       52.71
4 1748541      0.28         0.24       50.21      57.61      4.80      0.28         0.24       49.75      57.43      5.00      0.28         0.24       50.21
5 1748543      0.28         0.24       51.60      58.85      4.23      0.28         0.24       49.04      56.92      5.29      0.28         0.24       51.60
6 1748545      0.28         0.25       49.03      56.91      5.29      0.28         0.25       49.03      56.91      5.29      0.28         0.25       49.03
  TOT_NO4AVE TOT_WTDEP ACC_SILTAVE ACC_CLAYAVE ACC_SANDAVE CAT_SILTAVE CAT_CLAYAVE CAT_SANDAVE TOT_SILTAVE TOT_CLAYAVE TOT_SANDAVE
1      57.25      4.89       59.74       14.56       25.70       59.74       14.56       25.70       59.74       14.56       25.70
2      58.49      4.41       59.52       13.75       26.73       59.52       13.75       26.73       59.52       13.75       26.73
3      59.68      3.77       58.99       13.47       27.53       58.99       13.47       27.53       58.99       13.47       27.53
4      57.61      4.80       59.72       14.43       25.84       59.90       14.89       25.21       59.72       14.43       25.84
5      58.85      4.23       59.33       14.02       26.65       60.09       15.28       24.63       59.33       14.02       26.65
6      56.91      5.29       60.09       15.28       24.62       60.09       15.28       24.62       60.09       15.28       24.62
```

**Notes of `p1_statsgo_soil_df` target:**
- The final attributes I selected for TOT, ACC, CAT are `c("KFACT","KFACT_UP","NO10AVE",
                                                        "NO4AVE","SILTAVE","CLAYAVE","SANDAVE","WTDEP")` (see 
attr description in abstract and metadata for [layer attr](https://www.sciencebase.gov/catalog/item/5728decfe4b0b13d3918a9aa) and [texture attr](https://www.sciencebase.gov/catalog/item/5728dd46e4b0b13d3918a9a7)). Please let me know if there are any other variables we want

### CODE REVIEW

File: _targets.R [hunk 1]
```diff
@@ -1,10 +1,23 @@
...
 source("1_fetch.R")
 
+# Define minor HUCs (hydrologic unit codes) that make up the DRB to use in calls to dataRetrieval functions
+# Lower Delaware: 0204 subregion (for now, exclude New Jersey Coastal (https://water.usgs.gov/GIS/huc_name.html)
+drb_huc8s <- c("02040101","02040102","02040103","02040104","02040105","02040106",
+               "02040201","02040202","02040203","02040204","02040205",
+               "02040206","02040207")
+
+## nhd parent id 
+# pulled from https://www.sciencebase.gov/catalog/item/5728d6ace4b0b13d3918a992
```
Comment (lekoenig) [inline _targets.R:18]: Thanks for adding this data provenance statement!

File: 1_fetch.R [hunk 2]
```diff
@@ -46,9 +47,39 @@ p1_targets_list <- list(
...
-  )
-
+  ),
+  
+  # STATSGO SOIL Characteristics
+  ## get selected child items nhdv2 STATSGO Soil Characteristics
+  ## 1) Text attributes and 2) Layer attributes
+  tar_target(
+    p1_selected_statsgo_sbid_children,
+    item_list_children(sb_id = nhd_statsgo_parent_sbid) %>%
```
Comment (lekoenig) [inline 1_fetch.R:57]:
```suggestion
    sbtools::item_list_children(sb_id = nhd_statsgo_parent_sbid) %>% 
```

I see that you use this format in the next target, so for consistency I suggested the same here.

File: 1_fetch.R [hunk 3]
```diff
@@ -46,9 +47,39 @@ p1_targets_list <- list(
...
-
+  ),
+  
+  # STATSGO SOIL Characteristics
+  ## get selected child items nhdv2 STATSGO Soil Characteristics
+  ## 1) Text attributes and 2) Layer attributes
+  tar_target(
+    p1_selected_statsgo_sbid_children,
+    item_list_children(sb_id = nhd_statsgo_parent_sbid) %>% 
+      Filter(function(x){str_detect(x[['title']],'Text|Layer')},
```
Comment (lekoenig) [inline 1_fetch.R:58]: Huh, that's neat. I'm not familiar with the `Filter()` function. Is this in `purrr`?

File: 1_fetch.R [hunk 4]
```diff
@@ -46,9 +47,39 @@ p1_targets_list <- list(
...
+             .)
+    ),
   
+  ## download selected CONUS STATSGO datasets from Science base
+  tar_target(
+    p1_download_statsgo_text_layer_attr,
```
Comment (lekoenig) [inline 1_fetch.R:64]:
```suggestion
    p1_download_statsgo_text_layer_zip,
```

Minor suggestion for the target name to make it apparent what the expected output looks like. If you do opt to go with this change, you'll need to also update the downstream targets that reference `p1_download_statsgo_text_layer_attr`.

File: 1_fetch/src/sb_read_filter_by_comids.R [hunk 5]
```diff
@@ -0,0 +1,43 @@
+#' Function to filter sb datasets to given comids in AOI. Generates unzip if needed.  
+
+sb_read_filter_by_comids <- function(comid, data_path, sb_comid_col = 'COMID',
+                                     selected_cols_contains = NULL, cbind = TRUE){
+  
+  #' @description function to unzip, fetch and munge datasets from Sciencebase. Built for STATSGO but generalized for ther SB datasets
+  #' @param comid character string or vector of character strings containing the common identifier (COMID) of the desired flowline(s)
+  #' @param data_path folder path to files downloaded from sb
+  #' @param sb_comid_col str of the the column name from the comids. Default is 'COMID' as this is typically the name of this column from ds from SB0
```
Comment (lekoenig) [inline 1_fetch/src/sb_read_filter_by_comids.R:9]: What does `ds from SB0` refer to? Are those specific statsgo data files?

File: 1_fetch/src/sb_read_filter_by_comids.R [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+  data <- lapply(data_lst, function(x) read.csv(x, sep = ',') %>% 
+                   ### selecting only specified columns of param selected_cols_contains
+                   {if(!is.null(selected_cols_contains))
+                     select(.,contains(append(sb_comid_col,
+                                              selected_cols_contains))) else .} %>% 
+                   ### filter to comid in AOI
+                   filter(.data[[sb_comid_col]] %in% comid))
+  
+  # 3) cbind if true
+  if(cbind == TRUE){
```
Comment (lekoenig) [inline 1_fetch/src/sb_read_filter_by_comids.R:38]: Are there use cases you have in mind for this function where we would want `cbind == FALSE`?

**Review by lekoenig — CHANGES_REQUESTED** [review]
This is looking good, @msleckman! I have a few comments/suggestions so would be curious about your thoughts on those before we merge. 

One of my comments was to suggest we use a function from the inland salinity repo to download the sb files. Our patterns will probably be pretty similar to a lot of the data download/processing that we did in inland salinity. Some aspects of those functions might differ (e.g. because we don't have a VarsOfInterest table), but we should try to draw on and modify those functions whenever we can to save ourselves time.

File: 1_fetch.R (hunk 3 above)
Comment (msleckman) [reply 1_fetch.R:58]: Actually, I believe `Filter()` is base. can filter items of a vector. Sort of the same family as `Reduce()`.

File: 1_fetch/src/sb_read_filter_by_comids.R (hunk 5 above)
Comment (msleckman) [reply 1_fetch/src/sb_read_filter_by_comids.R:9]: not sure what i am refering to there 🤔, i think I mean `column in datasets from SB`. I'll update.

File: 1_fetch/src/sb_read_filter_by_comids.R (hunk 6 above)
Comment (msleckman) [reply 1_fetch/src/sb_read_filter_by_comids.R:38]: What I was envisioning here is that if we do not want to combine TOT, CAT, and ACC and keep the output as a list of distinct df.

File: _targets.R [hunk 7]
```diff
@@ -1,10 +1,23 @@
...
+# dir for selected datasets soil characteristics from nhd statsgo
+dir.create('1_fetch/out/statsgo', showWarnings = FALSE)
 
 source("1_fetch.R")
 
+# Define minor HUCs (hydrologic unit codes) that make up the DRB to use in calls to dataRetrieval functions
+# Lower Delaware: 0204 subregion (for now, exclude New Jersey Coastal (https://water.usgs.gov/GIS/huc_name.html)
+drb_huc8s <- c("02040101","02040102","02040103","02040104","02040105","02040106",
+               "02040201","02040202","02040203","02040204","02040205",
+               "02040206","02040207")
```
[the remark this answers is not in the document]
Comment (msleckman) [reply _targets.R:15]: I included it as we previously thought we would need. Ultimately not necessary.

File: 1_fetch.R [hunk 8]
```diff
@@ -46,9 +47,39 @@ p1_targets_list <- list(
...
+    lapply(p1_selected_statsgo_sbid_children,
+           function(x){sbtools::item_file_download(x$id,
+                                                   dest_dir = '1_fetch/out/statsgo',
+                                                   overwrite_file = TRUE)}
+           ) %>%
+      unlist(),
+    format = 'file'
+    ),
```
[the remark this answers is not in the document]
Comment (msleckman) [reply 1_fetch.R]: incorporated this suggestion https://github.com/USGS-R/drb-gw-hw-model-prep/pull/4/commits/d10a72d36ebc5cf746b1af675fb651dec8c3db5b

File: 1_fetch.R (hunk 4 above)
Comment (msleckman) [reply 1_fetch.R]: changed.

**Comment by msleckman** [comment]
Changes incorporated and ready for merge.

File: _targets.R (hunk 7 above)
[the remark this answers is not in the document]
Comment (lekoenig) [reply _targets.R:15]: I can't say for certain that we won't use them at some point, but if you're not using `drb_huc8s` in this PR I think it's safe to omit for now.

[the remark this answers is not in the document]
Comment (msleckman) [reply _targets.R]: Omitted. https://github.com/USGS-R/drb-gw-hw-model-prep/pull/4/commits/e65b65c354106774e1418bd7454f3ae7862b3e8b

**Review by lekoenig — APPROVED** [review]

**Comment by msleckman** [comment]
Pipeline runs successfully for me with latest merge from main. 

image

**Pull request merged**